### PR TITLE
feat: remove fileNames and refactor tests

### PR DESCRIPTION
**Pull request by dgggIoT**
**Issue #, if available:**
#51 
**Description of changes:**
Mainly remove all places that use fileNames, now we only use fileHash. 
**Why is this change necessary:**

**How was this change tested:**

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ x] Updated or added new unit tests
 - [ x] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/logmanager/model/ComponentLogConfiguration.java [hunk 1]
```diff
@@ -25,6 +25,5 @@ public class ComponentLogConfiguration {
     private Level minimumLogLevel = Level.INFO;
     private Long diskSpaceLimit;
     private boolean deleteLogFileAfterCloudUpload;
-    private boolean uploadToCloudWatch;
```
Comment (dgggIoT) [inline src/main/java/com/aws/greengrass/logmanager/model/ComponentLogConfiguration.java:28]: This is wrongly added in previous PR.

**Review by dgggIoT — COMMENTED** [review]
This PR passes all tests locally. 
However, recently, all PRs, including previous merged PRs, fail the github check. This seems like a framework issue.

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/logmanager/smallSpaceManagementPeriodicIntervalConfig.yaml [hunk 2]
```diff
@@ -10,7 +10,7 @@ services:
             minimumLogLevel: 'INFO'
             diskSpaceLimit: '105'
             diskSpaceLimitUnit: 'KB'
-            deleteLogFileAfterCloudUpload: 'true'
+            deleteLogFileAfterCloudUpload: 'false'
```
Comment (dgggIoT) [inline src/integrationtests/resources/com/aws/greengrass/integrationtests/logmanager/smallSpaceManagementPeriodicIntervalConfig.yaml:13]: This is only used for diskSpaceManagement Test, where files should not be processed.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerConfigTest.java [hunk 3]
```diff
@@ -95,27 +102,42 @@ static void setupKernel() throws InterruptedException,
         assertTrue(logManagerRunning.await(10, TimeUnit.SECONDS));
     }
 
-    @AfterAll
-    static void afterAll() {
+    @BeforeEach
+    void beforeEach(ExtensionContext context)
+            throws DeviceConfigurationException, URISyntaxException, IOException, InterruptedException {
+        ignoreExceptionOfType(context, TLSAuthException.class);
```
Comment (Nelsonochoam) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerConfigTest.java:108]: Is this really needed for all the tests - We should just add it to the tests when they throw that type of exception and we expect it. Otherwise this just hides some real errors

Comment (dgggIoT) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerConfigTest.java:108]: This test was designed not to process any files. But now the active file is processed.

File: src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java [hunk 4]
```diff
@@ -225,14 +224,14 @@ private boolean processLogLine(AtomicInteger totalBytesRead,
...
                                    StringBuilder data,
-                                   String fileName,
                                    String fileHash,
                                    long startPosition,
                                    String componentName,
                                    long currentPosition,
                                    long lastModified,
                                    LogFileGroup logFileGroup) {
         String dataStr = data.toString();
+        //System.out.println(dataStr);
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java:234]: remove

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 5]
```diff
@@ -78,7 +81,7 @@ public class LogManagerService extends PluginService {
     public static final String LOGS_UPLOADER_PERIODIC_UPDATE_INTERVAL_SEC = "periodicUploadIntervalSec";
     public static final String LOGS_UPLOADER_CONFIGURATION_TOPIC = "logsUploaderConfiguration";
     public static final String SYSTEM_LOGS_COMPONENT_NAME = "System";
-    public static final String PERSISTED_CURRENT_PROCESSING_FILE_NAME = "currentProcessingFileName";
+    public static final String PERSISTED_CURRENT_PROCESSING_FILE_HASH = "currentProcessingFileHash";
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:84]: What would happen when customer upgrades log manager from an older version?
1. Would the new version upload logs that's already uploaded by an older version?
2. How does log manager clean up the deprecated config nodes?

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 6]
```diff
@@ -675,7 +663,14 @@ private void processLogsAndUpload() throws InterruptedException {
         }
     }
 
+    public Runnable registerEventStatusListener(Consumer<EventType> callback) {
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:666]: Where is this function used?

Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:666]: In the Integ test. This is just for allowing test can track the status of service.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 5 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:84]: Thanks for reminding. This will be handled in the next PR.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 7]
```diff
@@ -116,6 +117,7 @@ public class LogManagerService extends PluginService {
     Map<String, ComponentLogConfiguration> componentLogConfigurations = new ConcurrentHashMap<>();
     @Getter
     private final CloudWatchLogsUploader uploader;
+    @Getter
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:120]: Why do we need a `Getter` on this one now?

Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:120]: Some tests needs to monitor provessor`s state.

Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:120]: Take my words back, some tests used to need it, but not anymore. Will remove it.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 8]
```diff
@@ -833,33 +828,39 @@ public void shutdown() throws InterruptedException {
...
-                    break;
                 case PERSISTED_CURRENT_PROCESSING_FILE_START_POSITION:
                     startPosition = Coerce.toLong(topic);
                     break;
                 case PERSISTED_CURRENT_PROCESSING_FILE_LAST_MODIFIED_TIME:
                     lastModifiedTime = Coerce.toLong(topic);
                     break;
+                case PERSISTED_CURRENT_PROCESSING_FILE_HASH:
+                    // This handles the upgrade scenario. ALl stored fileNames are transferred to the fileHash.
+                    String savedString = Coerce.toString(topic);
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:857]: Instead of deleting the old key on this release we should keep writing to both and but only read from the hash. In base someone upgrades and then downgrades at least it would keep working

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 9]
```diff
@@ -833,33 +828,39 @@ public void shutdown() throws InterruptedException {
...
         @JsonProperty(PERSISTED_CURRENT_PROCESSING_FILE_START_POSITION)
         private long startPosition;
         @JsonProperty(PERSISTED_CURRENT_PROCESSING_FILE_LAST_MODIFIED_TIME)
         private long lastModifiedTime;
+        @JsonProperty(PERSISTED_CURRENT_PROCESSING_FILE_HASH)
+        private String fileHash;
 
         public Map<String, Object> convertToMapOfObjects() {
             Map<String, Object> currentProcessingFileInformationMap = new HashMap<>();
-            currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_NAME, fileName);
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:845]: Lets avoid deleting this key and I think it is ok to keep the fileName here. Just tag it as deprecated and signal that it should be removed. I just want to avoid people upgrading and then downgrading from breaking

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 8 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:857]: Agree. I remove it first, and need to handle it in a separate PR.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 9 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:845]: Agree.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerTest.java [hunk 10]
```diff
@@ -337,120 +304,30 @@ void GIVEN_log_manager_in_errored_state_WHEN_restarted_THEN_logs_upload_is_reatt
...
             assertNotNull(request.logEvents());
-            assertEquals(50, request.logEvents().size());
-            assertEquals(51200, request.logEvents().stream().mapToLong(value -> value.message().length()).sum());
+            assertEquals(DEFAULT_LOG_LINE_IN_FILE * testFileNumber, request.logEvents().size());
+            assertEquals(DEFAULT_FILE_SIZE * testFileNumber,
+                    request.logEvents().stream().mapToLong(value -> value.message().length()).sum());
         }
-        File folder = tempDirectoryPath.toFile();
         Pattern logFileNamePattern = Pattern.compile("^integTestRandomLogFiles.log\\w*");
-        List<File> allFiles = new ArrayList<>();
```
Comment (Nelsonochoam) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerTest.java:345]: Love that this is getting nuked

File: src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java [hunk 11]
```diff
@@ -96,12 +97,14 @@ public void startup() {
 
     @Test
     void GIVEN_one_system_component_one_file_less_than_max_WHEN_merge_THEN_reads_entire_file(ExtensionContext ec)
-            throws URISyntaxException {
+            throws URISyntaxException, IOException, InvalidLogGroupException {
         ignoreExceptionOfType(ec, DateTimeParseException.class);
+        //LogFile logFile1 = createLogFileWithSize(getClass().getResource("testlogs2.log").toURI(), 1061);
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java:102]: Delete?

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 5 above)
Comment (Nelsonochoam) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:84]: I added this on another comment but I think we should keep this the old variable and still write to it just stop reading from it and read from the hash to avoid any potential breaking change add a @depreacted annotation and a comment to remove it later, that will derisk things in case of upgrade/downgrades

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 12]
```diff
@@ -78,7 +80,9 @@ public class LogManagerService extends PluginService {
     public static final String LOGS_UPLOADER_PERIODIC_UPDATE_INTERVAL_SEC = "periodicUploadIntervalSec";
     public static final String LOGS_UPLOADER_CONFIGURATION_TOPIC = "logsUploaderConfiguration";
     public static final String SYSTEM_LOGS_COMPONENT_NAME = "System";
+    //This is a deprecated value in the new version, but keep it here to avoid upgrade-downgrade issues.
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:83]: Specify that it is deprecated in versions greater than 2.2

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 13]
```diff
@@ -504,68 +506,57 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
+                    if (fileInformation.fileHash.equals(fileHash)) {
+                        currentProcessingLogFilePerComponent.remove(componentName);
+                    }
                 }
+            } else {
+                // Add the file to the current processing list for the component.
+                // Note: There should always be only 1 file which will be in progress at any given time.
+                CurrentProcessingFileInformation processingFileInformation =
+                        CurrentProcessingFileInformation.builder()
+                                .startPosition(cloudWatchAttemptLogFileInformation.getStartPosition()
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:549]: IMPORTANT -> Keep passing the file name here

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 14]
```diff
@@ -833,24 +828,30 @@ public void shutdown() throws InterruptedException {
...
             currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_START_POSITION, startPosition);
             currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_LAST_MODIFIED_TIME,
                     lastModifiedTime);
+            currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_HASH, fileHash);
             return currentProcessingFileInformationMap;
         }
 
         public void updateFromTopic(Topic topic) {
             switch (topic.getName()) {
+                //This is a deprecated value in the new version, but keep it here to avoid upgrade-downgrade issues.
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:854]: Specify that it is deprecated after 2.2

File: pom.xml [hunk 15]
```diff
@@ -442,6 +442,6 @@
         <maven.compiler.target>1.8</maven.compiler.target>
         <skipTests>false</skipTests>
         <groups></groups>
-        <logmanagerjar.name>LogManager</logmanagerjar.name>
+        <logmanagerjar.name>aws.greengrass.LogManager</logmanagerjar.name>
```
Comment (Nelsonochoam) [inline pom.xml:445]: Don't commit this change - it was good for testing but we don't need this to be part of this changes

File: recipe.json [hunk 16]
```diff
@@ -0,0 +1,27 @@
+{
+    "RecipeFormatVersion": "2020-01-25",
+    "ComponentName": "COMPONENT_NAME",
+    "ComponentVersion": "COMPONENT_VERSION",
+    "ComponentType": "aws.greengrass.plugin",
+    "ComponentDescription": "test for log manager",
+    "ComponentPublisher": "COMPONENT_PUBLISHER",
+    "ComponentDependencies": {
```
Comment (Nelsonochoam) [inline recipe.json:8]: Don't commit this file here - good for testing but we don't need to add this as part of these changes

File: recipe.yaml [hunk 17]
```diff
@@ -1,10 +0,0 @@
----
-RecipeFormatVersion: '2020-01-25'
-ComponentName: aws.greengrass.LogManager
-ComponentDescription: AWS Greengrass Log Manager
```
Comment (Nelsonochoam) [inline recipe.yaml:4]: Don't commit this change - this was only for testing your changes (revert this file)

File: src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/SpaceManagementTest.java [hunk 18]
```diff
@@ -134,26 +132,14 @@ void GIVEN_user_component_config_with_space_management_WHEN_space_exceeds_THEN_e
             }
         };
         try (AutoCloseable l = createCloseableLogListener(listener)) {
-            for (int i = 0; i < 14; i++) {
+            for (int i = 0; i < 15; i++) {
                 createTempFileAndWriteData(tempDirectoryPath, "integTestRandomLogFiles.log_", "");
             }
             assertTrue(cdl.await(60, TimeUnit.SECONDS), "5 files deleted");
```
Comment (Nelsonochoam) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/SpaceManagementTest.java:138]: Ugh.. this test is not great... it is checking for log messages to see if a file got deleted. Could we add a TODO on line 127 

```suggestion
          // TODO: We should not be checking for log messages to verify if a log file was deleted.
```

Also it is not verifying the active file is still present

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 19]
```diff
@@ -78,7 +80,10 @@ public class LogManagerService extends PluginService {
     public static final String LOGS_UPLOADER_PERIODIC_UPDATE_INTERVAL_SEC = "periodicUploadIntervalSec";
     public static final String LOGS_UPLOADER_CONFIGURATION_TOPIC = "logsUploaderConfiguration";
     public static final String SYSTEM_LOGS_COMPONENT_NAME = "System";
+    //This is deprecated value in versions greater than 2.2, but keep it here to avoid
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:83]:
```suggestion
    // @deprecated This is deprecated value in versions greater than 2.2, but keep it here to avoid
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 20]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                Optional<LogFile> activeFile = logFileGroup.getActiveFile();
-                if (activeFile.isPresent()) {
-                    isActiveFile = activeFile.get().fileEquals(file);
-                }
-            } catch (InvalidLogGroupException e) {
-                logger.atDebug().cause(e).log();
+        try {
+            LogFileGroup logFileGroup = attemptLogInformation.getLogFileGroup().syncDirectory();
+            if (!logFileGroup.getFile(fileHash).isPresent()) {
+                logger.atTrace().log("The fileHash does not exist in directory");
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:519]:
```suggestion
                logger.atTrace().kv("fileHash", fileHash).log("component", logFileGroup.getPattern()).log("File not found in directory");
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 21]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                //  We will grab the fileHash in the future.
-                Optional<LogFile> activeFile = logFileGroup.getActiveFile();
-                if (activeFile.isPresent()) {
-                    isActiveFile = activeFile.get().fileEquals(file);
-                }
-            } catch (InvalidLogGroupException e) {
-                logger.atDebug().cause(e).log();
+        try {
+            LogFileGroup logFileGroup = attemptLogInformation.getLogFileGroup().syncDirectory();
+            if (!logFileGroup.getFile(fileHash).isPresent()) {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:518]:
```suggestion
         LogFile logFile =  logFileGroup.getFile(fileHash);

         if (!file.isPresent()) {
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 22]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                        .get(componentName);
-                if (fileInformation.fileName.equals(fileName)) {
-                    currentProcessingLogFilePerComponent.remove(componentName);
+            LogFile file = logFileGroup.getFile(fileHash).get();
+            Optional<LogFile> activeFile = logFileGroup.getActiveFile();
+            boolean isActiveFile = false;
+            if (activeFile.isPresent()) {
+                isActiveFile = activeFile.get().fileEquals(file);
+            }
+            //This is deprecated value in versions greater than 2.2, but keep it here to avoid
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:528]:
```suggestion
            // @deprecated This is deprecated value in versions greater than 2.2, but keep it here to avoid
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 23]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
+            // upgrade-downgrade issues.
+            String fileName = activeFile.get().getAbsolutePath();
+            // If we have completely read the file, then we need add it to the completed files list and remove it
+            // it (if necessary) for the current processing list.
+            String componentName = attemptLogInformation.getComponentName();
+            if (!isActiveFile && file.length() == cloudWatchAttemptLogFileInformation.getBytesRead()
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:534]: Add a method to the LogFile returned by the log group that you can use to check if the file is the active one or not

```suggestion
            if (!logFile.isActive() && file.length() == cloudWatchAttemptLogFileInformation.getBytesRead()
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 24]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                    currentProcessingLogFilePerComponent.remove(componentName);
+            LogFile file = logFileGroup.getFile(fileHash).get();
+            Optional<LogFile> activeFile = logFileGroup.getActiveFile();
+            boolean isActiveFile = false;
+            if (activeFile.isPresent()) {
+                isActiveFile = activeFile.get().fileEquals(file);
+            }
+            //This is deprecated value in versions greater than 2.2, but keep it here to avoid
+            // upgrade-downgrade issues.
+            String fileName = activeFile.get().getAbsolutePath();
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:530]: ⚠️ CAREFUL with this one  - you are always making the fileName be the active file name. That is not how it used to work. This fileName was from the parameter that was being passed so this should basically be the `logFile.get().getAbsolutePath()`

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 25]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
+                    CurrentProcessingFileInformation fileInformation = currentProcessingLogFilePerComponent
+                            .get(componentName);
+                    if (fileInformation.fileHash.equals(fileHash)) {
+                        currentProcessingLogFilePerComponent.remove(componentName);
+                    }
                 }
+            } else {
+                // Add the file to the current processing list for the component.
+                // Note: There should always be only 1 file which will be in progress at any given time.
+                //The fileName is deprecated value in versions greater than 2.2, but keep it here to avoid
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:550]:
```suggestion
                // @deprecated The fileName is deprecated value in versions greater than 2.2, but keep it here to avoid
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 26]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                            .fileName(fileName)
-                            .startPosition(cloudWatchAttemptLogFileInformation.getStartPosition()
-                                    + cloudWatchAttemptLogFileInformation.getBytesRead())
-                            .lastModifiedTime(cloudWatchAttemptLogFileInformation.getLastModifiedTime())
-                            .build();
-            currentProcessingLogFilePerComponent.put(attemptLogInformation.getComponentName(),
-                    processingFileInformation);
+        } catch (InvalidLogGroupException e) {
+            logger.atDebug().cause(e).log();
+            return;
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:565]: This return is not needed.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 27]
```diff
@@ -504,68 +506,63 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                    CurrentProcessingFileInformation.builder()
-                            .fileName(fileName)
-                            .startPosition(cloudWatchAttemptLogFileInformation.getStartPosition()
-                                    + cloudWatchAttemptLogFileInformation.getBytesRead())
-                            .lastModifiedTime(cloudWatchAttemptLogFileInformation.getLastModifiedTime())
-                            .build();
-            currentProcessingLogFilePerComponent.put(attemptLogInformation.getComponentName(),
-                    processingFileInformation);
+        } catch (InvalidLogGroupException e) {
+            logger.atDebug().cause(e).log();
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:564]:
```suggestion
            logger.atDebug().cause(e).log("Invalid log group");
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 28]
```diff
@@ -628,8 +625,7 @@ private void processLogsAndUpload() throws InterruptedException {
                             if (componentCurrentProcessingLogFile.containsKey(componentName)) {
                                 CurrentProcessingFileInformation processingFileInformation =
                                         componentCurrentProcessingLogFile.get(componentName);
-                                if (processingFileInformation.fileName.equals(file.getAbsolutePath())
-                                        && processingFileInformation.lastModifiedTime == file.lastModified()) {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:632]: Why is this check for lastModifiedTime not needed anymore? - Is it because of the same name rotation?

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 29]
```diff
@@ -658,18 +654,27 @@ private void processLogsAndUpload() throws InterruptedException {
...
                 isCurrentlyUploading.set(false);
             } else {
                 unitsOfWork.forEach((unit) -> {
                     CloudWatchAttempt cloudWatchAttempt = logsProcessor.processLogFiles(unit);
                     uploader.upload(cloudWatchAttempt, 1);
                 });
+                emitEventStatus(EventType.LOG_GROUP_PROCESSED);
             }
+            // after handle one cycle, we sleep for interval to avoid seamless scanning and processing next cycle.
+            TimeUnit.SECONDS.sleep(periodicUpdateIntervalSec);
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:666]: ⚠️ This might not be what we want - it concerns me a bit that tests pass with this (meaning we don't have a test for this) - This forces the thread to always sleep after each cycle if a client sets it to 20 minutes then we will still have a similar effect to the one we have today where active log files take forever to be uploaded.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 30]
```diff
@@ -658,18 +654,27 @@ private void processLogsAndUpload() throws InterruptedException {
...
 
             if (unitsOfWork.isEmpty()) {
-                TimeUnit.SECONDS.sleep(periodicUpdateIntervalSec);
                 isCurrentlyUploading.set(false);
             } else {
                 unitsOfWork.forEach((unit) -> {
                     CloudWatchAttempt cloudWatchAttempt = logsProcessor.processLogFiles(unit);
                     uploader.upload(cloudWatchAttempt, 1);
                 });
+                emitEventStatus(EventType.LOG_GROUP_PROCESSED);
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:663]: Maybe time to change the event name given each cycle might process multiple log groups on the `while` loop

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 31]
```diff
@@ -827,24 +832,33 @@ public void shutdown() throws InterruptedException {
...
         @JsonProperty(PERSISTED_CURRENT_PROCESSING_FILE_START_POSITION)
         private long startPosition;
         @JsonProperty(PERSISTED_CURRENT_PROCESSING_FILE_LAST_MODIFIED_TIME)
         private long lastModifiedTime;
+        @JsonProperty(PERSISTED_CURRENT_PROCESSING_FILE_HASH)
+        private String fileHash;
 
         public Map<String, Object> convertToMapOfObjects() {
             Map<String, Object> currentProcessingFileInformationMap = new HashMap<>();
+            //This is deprecated value in versions greater than 2.2, but keep it here to avoid
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:848]:
```suggestion
            // @deprecated This is deprecated value in versions greater than 2.2, but keep it here to avoid
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 32]
```diff
@@ -827,24 +832,33 @@ public void shutdown() throws InterruptedException {
...
             currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_START_POSITION, startPosition);
             currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_LAST_MODIFIED_TIME,
                     lastModifiedTime);
+            currentProcessingFileInformationMap.put(PERSISTED_CURRENT_PROCESSING_FILE_HASH, fileHash);
             return currentProcessingFileInformationMap;
         }
 
         public void updateFromTopic(Topic topic) {
             switch (topic.getName()) {
+                //This is deprecated value in versions greater than 2.2, but keep it here to avoid
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:860]:
```suggestion
                // @deprecated This is deprecated value in versions greater than 2.2, but keep it here to avoid
```

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 28 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:632]: My guess: it was just a double confirmation.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 21 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:518]: Refactoring a little bit, please see next commit

File: .gitignore [hunk 33]
```diff
@@ -10,3 +10,7 @@ coverage
 *.iml
 bin
 greengrass*.log
+gdk-config.json
```
Comment (Nelsonochoam) [inline .gitignore:13]: Just don't place this files here. Revert this

**Review by Nelsonochoam — APPROVED** [review]
I am approving this knowing that a lot of the tests need to be refactored given they are hard to follow. @dgggIoT tested it manually to ensure the existing functionality still worked and that the active file was being uploaded. There are still a few follow up PRs that are needed and we should write or refactor a few more tests.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 34]
```diff
@@ -504,68 +505,58 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                if (activeFile.isPresent()) {
-                    isActiveFile = activeFile.get().fileEquals(file);
-                }
-            } catch (InvalidLogGroupException e) {
-                logger.atDebug().cause(e).log();
+        try {
+            LogFileGroup logFileGroup = attemptLogInformation.getLogFileGroup().syncDirectory();
+            if (!logFileGroup.isHashExist(fileHash)) {
+                logger.atTrace().kv("fileHash", fileHash).log("component",
+                        logFileGroup.getFilePattern(), "File not found in directory");
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:519]: When would we reach this log? A file being deleted?

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 35]
```diff
@@ -504,68 +505,58 @@ private void handleCloudWatchAttemptStatus(CloudWatchAttempt cloudWatchAttempt)
...
-                if (fileInformation.fileName.equals(fileName)) {
-                    currentProcessingLogFilePerComponent.remove(componentName);
+            LogFile file = logFileGroup.getFile(fileHash);
+            // @deprecated  This is deprecated value in versions greater than 2.2, but keep it here to avoid
+            // upgrade-downgrade issues.
+            String fileName = file.getAbsolutePath();
+            // If we have completely read the file, then we need add it to the completed files list and remove it
+            // it (if necessary) for the current processing list.
+            String componentName = attemptLogInformation.getComponentName();
+            if (!logFileGroup.isActiveFile(file) && file.length() == cloudWatchAttemptLogFileInformation.getBytesRead()
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:529]: nit: maybe change `==` to `<=`? This seems like a very delicate logic. If somehow bytes calculating went wrong then a file could be stuck in processing forever.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java [hunk 36]
```diff
@@ -657,19 +647,25 @@ private void processLogsAndUpload() throws InterruptedException {
...
-                    CloudWatchAttempt cloudWatchAttempt = logsProcessor.processLogFiles(unit);
-                    uploader.upload(cloudWatchAttempt, 1);
-                });
-            }
+            unitsOfWork.forEach((unit) -> {
+                CloudWatchAttempt cloudWatchAttempt = logsProcessor.processLogFiles(unit);
+                uploader.upload(cloudWatchAttempt, 1);
+            });
+            isCurrentlyUploading.set(false);
+            emitEventStatus(EventType.ALL_COMPONENTS_PROCESSED);
```
[the remark this answers is not in the document]
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:655]: Another potential alternative would be asserting on log manager's log.

**Review by junfuchen99 — APPROVED** [review]
I'm fine with approving this PR to unblock your development. Overall the current code s(both business logic and tests) are still very difficult to read and I hope you can improve it in your follow-up PRs.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 34 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:519]: when the file is under classification of processing or completed.

File: src/main/java/com/aws/greengrass/logmanager/LogManagerService.java (hunk 35 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/LogManagerService.java:529]: Will add it in next PR.
